tasty/skyspark/process_graphs.py

```diff
@@ -0,0 +1,256 @@
+        """
+
+        # get the source shapes directory based on the given schema and create a list of all the json files in that directory
+        source_shapes_schema_dir = os.path.join(source_shapes_dir, self.schema.lower())
+        files = [os.path.join(source_shapes_schema_dir, f) for f in
+                 os.listdir(source_shapes_schema_dir) if f.endswith('.json')]
+
+        valid_tags = []
+        valid_tags_ns = []
+
+        # go through schema files and extract valid tags
+        for file in files:
+            # open file and read in json to python dict
+            with open(file, 'r') as f:
+                filedata = json.loads(f.read())
+                # for each shape
+                for shape in filedata['shapes']:
+                    # add tags if not already added to the tags list
+                    if 'tags' in shape:
+                        for tag in shape['tags']:
+                            if tag not in valid_tags:
+                                valid_tags.append(tag)
+                    # add custom tags if not already added to the tags list
+                    if 'tags-custom' in shape:
+                        for tag in shape['tags-custom']:
+                            if tag not in valid_tags:
+                                valid_tags.append(tag)
+
+        print("...generated tag list")
+        print("...adding namespaces")
+        # sort tags list
+        valid_tags = sorted(valid_tags)
+
+        # add namespaces to all valid tags
+        for tag in valid_tags:
+            tag_ns = tg.get_namespaced_term(self.ontology_graph, tag)
+            # take care of custom tags
+            if tag_ns is False:
+                tag_ns = PHCUSTOM[tag]
+            valid_tags_ns.append(tag_ns)
+
+        tags_dict = {
+            'plain': valid_tags,
+            'namespaced': valid_tags_ns
+        }
+
+        return tags_dict
+
+    def remove_invalid_tags(self, data_graph):
+        """
+        This method retrieves the list of valid tags, then checks the tags of each point in the given rdf grpah against this list.
+        If the tags are not in the list of valid tags, they are removed from the graph
+
+        :param data_graph: the rdf (instance data) graph to process
+        """
+
+        # get the list of valid tags
+        valid_tags = self.get_valid_tags()
+        valid_tags_ns = valid_tags['namespaced']
+
+        # Iterate over all points in the graph (anything with an "equipRef" tag)
+        for s1, p1, o1 in data_graph.triples((None, tc.PHIOT_3_9_10["equipRef"], None)):
+            print(f"...processing node: \t{s1}")
+            # iterate over each tag
+            for s, p, o in data_graph.triples((s1, tc.PH_3_9_10["hasTag"], None)):
+                # if the tag is not in the list of valid tags, remove it
+                if o not in valid_tags_ns:
+                    data_graph.remove((s, p, o))
```

Comment:
Does this work if the class was instantiated with `version=tc.v3_9_9` above? See my general comment about generalizing schema and version where possible.